src/cargo/sources/git/utils.rs

```diff
@@ -180,12 +179,48 @@ impl GitDatabase {
     }
 
     pub fn resolve(&self, r: &GitReference) -> CargoResult<git2::Oid> {
-        r.resolve(&self.repo)
+        r.resolve(&self.repo, true)
     }
 }
 
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ShalowDataBlob<'a> {
+    tree: &'a str,
+    etag: &'a str,
+}
+
+#[test]
+fn check_with_git_hub() {
+    panic!(r#"nonstandard shallow clone may be worse than a full check out. 
+    This test is here to make sure we do not merge until we have official signoff from GitHub"#)
+}
+
 impl GitReference {
-    pub fn resolve(&self, repo: &git2::Repository) -> CargoResult<git2::Oid> {
+    pub fn resolve(&self, repo: &git2::Repository, tree: bool) -> CargoResult<git2::Oid> {
+        // Check if Cargo has done a nonstandard shallow clone
+        if let Some(reference) = repo
+            .find_reference(
+                &(format!(
+                    "refs/cargo-{}",
+                    serde_json::to_string(self).expect("why cant we make json of this")
+                )),
+            )
+            .ok()
+            .and_then(|re| {
+                let blob = re.peel_to_blob().ok()?;
+                let shalow_data: ShalowDataBlob<'_> =
+                    serde_json::from_slice(blob.content()).ok()?;
+                let id = if tree {
+                    shalow_data.tree
+                } else {
+                    shalow_data.etag
+                };
+                Some(id.parse::<Oid>().ok()?)
+            })
+        {
+            return Ok(reference);
```

Comment:
I'm concerned over the brittleness of this API.

If `tree` is true, it will resolve to a tree if and only if we are in the fast path.   All (transitive) callers need to know this is just a generic object id and not specifically a commit id (which is an easy assumption to make).

In my ideal world
- We would have separate `resolve_to_object` and `resolve_to_commit` functions.
  - `object` is more accurate than `tree` as we can't assume its a tree
  - I feel like we should make this more obvious in the caller than just using a `true` or `false`
  - separate functions as the difference in which to use is a design-time decision and not a runtime decision
- `resolve_to_object` would not do any `peel`s, making it more obvious when someone misuses the API

My main concern with this ideal is how much code would be duplicated.  I guess its not so bad if
- We have a `_resolve_to_object` function that is just the old code path without peels.  ``resolve_to_commit` would call it and do a peel.
- We have a `_find_shallow_blob` function that just gets the blob and each of the top-level functions extract the relevant field from that.

Thoughts?

Reply:
I completely agree with your point that this is a compile time distinction. I will rework the interface to match that (when I next have time to code). I think the two relevant functions are `resolve_to_tree` (which always points to a tree that git has on disk) and `resolve_to_commit` (which will always point to a commit, but may not be something git has on disk).